ext/bcmath/libbcmath/src/round.c

```diff
@@ -33,10 +33,50 @@ void bc_round(bc_num num, zend_long precision, zend_long mode, bc_num *result)
 	* - If the fractional part ends with zeros, the zeros are omitted and the number of digits in num is reduced.
 	*   Meaning we might end up in the previous case.
 	*/
+
+	/* e.g. value is 0.1 and precision is -3, ret is 0 or 1000  */
 	if (precision < 0 && num->n_len < (size_t) (-(precision + Z_L(1))) + 1) {
-		*result = bc_copy_num(BCG(_zero_));
+		switch (mode) {
+			case PHP_ROUND_HALF_UP:
+			case PHP_ROUND_HALF_DOWN:
+			case PHP_ROUND_HALF_EVEN:
+			case PHP_ROUND_HALF_ODD:
+			case PHP_ROUND_TOWARD_ZERO:
+				*result = bc_copy_num(BCG(_zero_));
+				return;
+
+			case PHP_ROUND_CEILING:
+				if (num->n_sign == MINUS) {
+					*result = bc_copy_num(BCG(_zero_));
+					return;
+				}
+				break;
+
+			case PHP_ROUND_FLOOR:
+				if (num->n_sign == PLUS) {
+					*result = bc_copy_num(BCG(_zero_));
+					return;
+				}
+				break;
+
+			case PHP_ROUND_AWAY_FROM_ZERO:
+				break;
+
+			EMPTY_SWITCH_DEFAULT_CASE()
+		}
+
+		if (bc_is_zero(num)) {
+			*result = bc_copy_num(BCG(_zero_));
+			return;
+		}
+
+		/* If precision is -3, it becomes 1000. */
+		*result = bc_new_num(-precision + 1, 0);
```

Comment:
This line will break under UBSAN for the following code:
```php
<?php
echo bcround("0.01", PHP_INT_MIN, RoundingMode::AwayFromZero);
```

This gives:
```
/run/media/niels/MoreData/php-8.4/ext/bcmath/libbcmath/src/round.c:74:13: runtime error: negation of -9223372036854775808 cannot be represented in type 'zend_long' (aka 'long'); cast to an unsigned type to negate this value to itself
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /run/media/niels/MoreData/php-8.4/ext/bcmath/libbcmath/src/round.c:74:13 
```

But since this is an extreme edge case that won't be hit in practice, I think it suffices here to just return 0 as a sentinel and throw an exception.

Reply:
@nielsdos 
Nice point, thanks!

The first argument is `size_t`, so how about this?

```
if (UNEXPECTED(precision == ZEND_LONG_MIN)) {
    *result = bc_new_num((size_t) ZEND_LONG_MAX + 2, 0);
}
```

Reply:
done!